File: C++/2331. Evaluate Boolean Binary Tree.cpp
class Solution {
private:
    bool inorder(TreeNode* root){
        if(root->left == NULL && root->right==NULL){
            return root->val;
        }
        else{
            if(root->val==2){
                return inorder(root->left) || inorder(root->right);
            }
            return inorder(root->left) && inorder(root->right);
        }
    }
public:
    bool evaluateTree(TreeNode* root) {
        return inorder(root);
    }
};

File: JAVA/2331. Evaluate Boolean Binary Tree.java
class Solution {
    public boolean evaluateTree(TreeNode root) {
         if(root == null) return false;        
        int rootValue = root.val;        
        boolean leftValue = evaluateTree(root.left);
        boolean rightValue = evaluateTree(root.right);        
        if(rootValue == 0) return false;
        if(rootValue == 1) return true;        
        if(rootValue == 2) return leftValue || rightValue;
        if(rootValue == 3) return leftValue && rightValue;        
        return false;
    }
}
